src/Facet/FacetContainer.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace Wikibase\DataModel\Facet;
+
+/**
+ * Interface for attaching and accessing facets (aka "roles objects" or "extensions").
+ *
+ * This interface provides the basis of an implementation of the Role Object Pattern, see
+ * <http://c2.com/cgi/wiki?RoleObjectPattern>.
+ *
+ * @since 5.0
+ *
+ * @licence GNU GPL v2+
+ * @author Daniel Kinzler
+ */
+interface FacetContainer {
+
+	/**
+	 * Determines whether the given facet is supported.
+	 *
+	 * @param string $name
+	 *
+	 * @return boolean
+	 */
+	public function hasFacet( $name );
+
+	/**
+	 * Returns a list of supported facet names.
+	 *
+	 * @return string[]
+	 */
+	public function listFacets();
+
+	/**
+	 * Returns a facet object.
+	 *
+	 * @param string $name
+	 * @param string|null $type The desired type
```

Comment:
1. What "type"? We are using this term for a lot of different things. I honestly have no idea which one is meant here.
2. "Desired" how?
3. What is supposed to happen when this is null?